README.md
# Cardinal Engine(Core) [B]

Что нужно для запуска сайта на этом движке?

Загрузить файлы на сервер. Зайти на сайт и следовать инструкциям установщика.
 
Далее идёт краткий обзор функционала с примерами его работы предоставляемый Cardinal Engine. Однако если Вы человек разбирающийся - Вы сможете легко и не принуждённо рассмотреть *ядро* самостоятельно для получения максимально возможного функционала.
 
## Что касается работы с базой данных.

#### Простое получение данных(одна запись):
```php
$row = db::doquery("SELECT * FROM news WHERE id = ".intval($_GET['id']));
```
#### Простое получение данных(все записи в таблице):
```php
$rows = db::doquery("SELECT * FROM news ORDER BY id DESC", true);
while($row = db::fetch_assoc()) {
...
}
```

## Что касается работы с шаблонизатором.

#### Добавление данных в шаблонизатор(одна запись):
```php
templates::assing_var("is_view", "1");
```
#### Добавление данных в шаблонизатор(множество записей):
```php
templates::assing_vars(array(
"is_view1" => "1",
"is_view2" => "2",
));
```
#### Добавление данных в шаблонизатор(множество записей циклом):
```php
for($i=0;$i<10;$i++) {
  templates::assing_vars(array(
  "is_view1" => "1",
  "is_view2" => "2",
  ), "news", "news".$i);
}
```

## Что касается, непосредственно работы с шаблоном.

#### Условия:
```php
[if 1==1]true[/if]
```
```php
[if 1==2]true[else]false[/if]
```
```php
[if 1==1]true[/if 1==1]
```
```php
[if 1==2]true[else 1==2]false[/if 1==2]
```
Условия могут быть вложенны только при условии их чёткого заверешния
```php
[if 1==2]
	true
[else 1==2]
	[if 1==1]true[else 1==1]false[/if 1==1]
[/if 1==2]
```

#### Циклы:
```php
[foreach block=news]
 <h1>{news.is_view1}</h1>
 {news.is_view2}
[/foreach]
```
```php
[foreach block=news]
 <h1>{news.is_view1}</h1>
 {news.is_view2}
[/foreach news]
```
##### !Внимание! Внутри цикла применяется отличная от обычных условий конструкция со схожей логикой:
```php
[foreach block=news]
[foreachif {news.is_view1}==1]
 <h1>{news.is_view1}</h1>[/foreachif]
 {news.is_view2}
[/foreach]
```

#Если у Вас есть какие-то вопросы - направляем их по-адресу: [email]
А так-же - следите за обновлениями.


[email]:mailto:killer-server@mail.ru


changelog.txt
Обозначения
~~~~~~~~~~~
[!] Нововведение или важное изменение/New features or important change
[+] Добавлено/New functions
[~] Изменено/Changed functions
[%] Исправление-багфикс/Bugfixes
[@] Эта информация будет интересна только админам и/или разработчикам
    This information is only for administrators and/or developers

Cardinal Engine Release 1
~~~~~~~~~~~~~~~~~~~~~~~~~~~
[!] ВНИМАНИЕ! Требуется версия PHP >= 5.3.2

2015-09-30 13:06:10 1.25.6-rc1
[+] Добавлен changelog для собственной навигации по-изменениям, а так-же для тех, кто хочет следить за развитием данного проекта
[+] Добавлена SEO-оптимизация для главной страницы
[+] Добавление авторского права в некоторые файлы, а со временем - указание такового во всех остальных

[@] Добавлена возможность подключать css-стили и/или js-скрипты из модулей как при помощи подключения файла, так и при помощи кода
[@] Добавлена возможность задачи необходимого количества страниц в "пагинаторе"
[@] Улучшение для большей гибкости системы SEO-оптимизации

[!] Добавлена константа {THEME} для подключаемых файлов

[%] Исправлена ошибка, не позволяющая подключить язык в модулях, которые загружаются сразу путём отключения подгрузки из базы данных


2015-10-03 9:59:12 1.25.6-rc2
[~] Исправлен README.md
[~] Измененна функция шаблона. Изменения коснулись вывода META-тегов и информации о логики работы шаблонов в функции
[~] Изменена функция получения случайного числа, теперь есть возможность получить случайное число от 0 до максимально возможного
[~] Убрана возможность удалить учётную запись, с которой сидит администратор

[+] Инсталятор с предупреждением о необходимости прав доступа к папкам
[+] Добавлен класс парсинга данных
[+] Добавлена проверка важных обновлений для ветки trunk, что в будущем - перекачюет и на ветку master. Информация о обновлениях - будет отображаться на главной странице панели управления.


2015-10-07 17:50:38 1.25.6-rc3
[+] Добавлен копирайт в некоторых файлах
[+] Добавлена возможность переподключения "ядра" без ошибок констант
[+] Добавлена поддержка "кук" для версии 7.0.0 и выше
[+] Добавлен выбор "драйвера" при установке
[+] Добавлено предупреждение о необходимости MbString
[+] Добавленна поддержка получения случайного числа для версий 7.0.0 как выше, так и ниже
[~] Исправленна ошибка при отключенной базе данных - вызов боковой панели
[~] Убрана лишнее наследование кеша и переписанно на статичный вызов
[~] Исправленны ссылки в пагинаторе возвращающие данные по-совершенно разным ключам
[~] Переработан шаблонизатор с учётом прямой и обратной совместимости с php 7.0.0 выше и ниже
[~] Исправленно использование глобальных переменных для доступа к уровню доступа пользователя в классе пользовательских уровней
[~] Переработана работа с базой данных, теперь имеется возможность как переключения, так и внедрения новых "драйверов" работы с базой данных при условии соблюдения должной иерархии
[%] Убраны лишние тестовые переменные "ноги" которых тянутся ещё с седых времён



install.php
<?php
/*
 *
 * @version 2015-10-07 17:50:38 1.25.6-rc3
 * @copyright 2014-2015 KilleR for Cardinal Engine
 *
 * Version Engine: 1.25.6-rc3
 * Version File: 2
 *
 * 2.1
 * add sql query for support updater
 * 2.2
 * add support stop installing without need modules in php
 * 2.3
 * add support "drivers" - submodules for database
 *
*/
define("IS_CORE", true);
define("IS_INSTALLER", true);
require_once("core.php");
if(isset($_GET['done'])) {
	templates::assign_vars(array("page" => "4"));
	echo templates::view(templates::complited_assing_vars("install", null, ""));
	die();
}
if(!isset($_SERVER['SERVER_NAME']) || (isset($_SERVER['HTTP_HOST']) && $_SERVER['HTTP_HOST']=="localhost") || empty($_SERVER['SERVER_NAME'])) {
	templates::assign_vars(array("page" => "error_server"));
	echo templates::view(templates::complited_assing_vars("install", null, ""));
	die();
}

if(sizeof($_POST)==0||sizeof($_POST)==1||sizeof($_POST)==2) {
	if(sizeof($_POST)==0) {
		templates::assign_vars(array("page" => "1"));
	} else if(sizeof($_POST)==1) {
		$cache = (get_chmod(ROOT_PATH."core/cache/")=="0777" ? "green":"red");
		$system_cache = (get_chmod(ROOT_PATH."core/cache/system/")=="0777" ? "green":"red");
		$mb = (function_exists('mb_detect_encoding') ? "green" : "red");
		if($cache=="red"||$system_cache=="red"||$mb=="red") {
			templates::assign_var("is_stop", "1");
		} else {
			templates::assign_var("is_stop", "0");
		}
		templates::assign_vars(array("page" => "2", "cache" => $cache, "system_cache" => $system_cache, "mb" => $mb));
	} else {
		templates::assign_vars(array("page" => "3", "SERNAME" => getenv('SERVER_NAME')));

	}
	$driver = ROOT_PATH."core/class/system/drivers/";
	$dirs = read_dir($driver, ".php");
	sort($dirs);
	for($i=0;$i<sizeof($dirs);$i++) {
		if($dirs[$i]=="index.php"||$dirs[$i]=="DriverParam.php"||$dirs[$i]=="drivers.php") {
			continue;
		}
		include_once($driver.$dirs[$i]);
		$dr_subname = str_replace(".php", "", $dirs[$i]);
		if(!class_exists($dr_subname)) {
			continue;
		}
		$dr_name = $dr_subname::$subname;
		templates::assign_vars(array("name" => $dr_subname, "value" => $dr_name), "drivers", "driver".$i);
	}
	echo templates::view(templates::complited_assing_vars("install", null, ""));
	die();
}

modules::manifest_set(array('functions', "create_pass"), "create_pass_install");
function create_pass_install($pass) {
	return sha1($pass);
}

$db_host = saves($_POST['db_host'], true);
$db_port = saves($_POST['db_port'], true);
$db_user = saves($_POST['db_user'], true);
$db_pass = saves($_POST['db_pass'], true);
$db_db = saves($_POST['db_db'], true);
$db_driver = saves($_POST['db_driver'], true);

db::changeDriver($db_driver);
db::OpenDriver();
if(!db::check_connect($db_host, $db_user, $db_pass)) {
	templates::assign_vars(array("page" => "error"));
	echo templates::view(templates::complited_assing_vars("install", null, ""));
	die();
}

$SQL = array();
$SQL[] = "DROP TABLE IF EXISTS `config`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `config` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `config_name` varchar(255) NOT NULL,
  `config_value` text NOT NULL,
  PRIMARY KEY (`id`),
  KEY `config_name` (`config_name`),
  FULLTEXT KEY `config_value` (`config_value`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";
$SQL[] = "INSERT INTO `config` SET `config_name` = \"db_version\", `config_value` = \"".VERSION."\"";

$SQL[] = "DROP TABLE IF EXISTS `comments`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `comments` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `u_id` varchar(255) NOT NULL,
  `added` varchar(255) NOT NULL,
  `ip` varchar(255) NOT NULL,
  `type` enum('catalog','salons') NOT NULL DEFAULT 'catalog',
  `time` int(11) NOT NULL,
  `comment` text NOT NULL,
  `parent_id` varchar(255) NOT NULL,
  `level` int(11) NOT NULL,
  `user_agent` varchar(255) NOT NULL,
  `email` varchar(255) NOT NULL,
  `guest` varchar(255) NOT NULL,
  `mod` enum('yes','no') NOT NULL DEFAULT 'no',
  `spam` enum('yes','no') NOT NULL DEFAULT 'no',
  PRIMARY KEY (`id`),
  FULLTEXT `added` (`added`),
  KEY `ip` (`ip`),
  KEY `data` (`u_id`),
  KEY `others` (`type`,`time`,`parent_id`,`level`),
  KEY `agent` (`user_agent`),
  KEY `spam` (`spam`),
  FULLTEXT KEY `email` (`email`),
  FULLTEXT KEY `guest` (`guest`),
  FULLTEXT KEY `comment` (`comment`)
) ENGINE=MyISAM  DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `error_log`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `error_log` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `times` int(11) NOT NULL,
  `ip` varchar(255) NOT NULL,
  `exception_type` varchar(255) NOT NULL,
  `message` varchar(255) NOT NULL,
  `filename` varchar(255) NOT NULL,
  `line` int(11) NOT NULL,
  `trace_string` longtext NOT NULL,
  `request_state` longtext NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM  DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `hackers`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `hackers` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `ip` varchar(255) NOT NULL,
  `page` text NOT NULL,
  `get` text NOT NULL,
  `post` text NOT NULL,
  `activ` enum('yes','no') NOT NULL DEFAULT 'yes',
  `referer` text NOT NULL,
  PRIMARY KEY (`id`),
  KEY `data` (`ip`,`activ`),
  FULLTEXT KEY `page` (`page`),
  FULLTEXT KEY `get` (`get`),
  FULLTEXT KEY `post` (`post`),
  FULLTEXT KEY `referer` (`referer`)
) ENGINE=MyISAM  DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `lang`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `lang` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `lang` varchar(255) NOT NULL,
  `orig` text NOT NULL,
  `translate` text NOT NULL,
  PRIMARY KEY (`id`),
  KEY `lang` (`lang`),
  FULLTEXT KEY `orig` (`orig`),
  FULLTEXT KEY `translate` (`translate`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `menu`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `menu` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL,
  `data` text NOT NULL,
  `position` int(11) NOT NULL,
  `menu` varchar(255) NOT NULL,
  `activ` enum('yes','no') NOT NULL DEFAULT 'yes',
  PRIMARY KEY (`id`),
  KEY `others` (`name`,`position`,`activ`),
  KEY `menu` (`menu`),
  FULLTEXT KEY `content` (`data`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `modules`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `modules` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `page` varchar(255) NOT NULL,
  `module` varchar(255) NOT NULL,
  `method` varchar(255) NOT NULL,
  `param` longtext,
  `activ` enum('yes','no') NOT NULL DEFAULT 'yes',
  `tpl` longtext,
  PRIMARY KEY (`id`),
  KEY `page` (`page`),
  KEY `modules` (`module`),
  KEY `method` (`method`),
  KEY `activ` (`activ`),
  FULLTEXT KEY `param` (`param`),
  FULLTEXT KEY `tpl` (`tpl`)
) ENGINE=MyISAM  DEFAULT CHARSET=utf8 DELAY_KEY_WRITE=1 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `users`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `users` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `username` varchar(255) NOT NULL,
  `alt_name` varchar(255) NOT NULL,
  `pass` varchar(255) NOT NULL,
  `admin_pass` varchar(255) NOT NULL,
  `light` varchar(255) NOT NULL,
  `level` int(11) NOT NULL,
  `avatar` varchar(255) NOT NULL,
  `email` varchar(255) NOT NULL,
  `reg_ip` varchar(255) NOT NULL,
  `last_ip` varchar(255) NOT NULL,
  `time_reg` int(11) NOT NULL,
  `last_activ` int(11) NOT NULL,
  `activ` enum('yes','no') NOT NULL DEFAULT 'yes',
  PRIMARY KEY (`id`),
  FULLTEXT `user` (`username`),
  FULLTEXT `pass` (`pass`, `admin_pass`),
  FULLTEXT `light` (`light`),
  KEY `level` (`level`),
  FULLTEXT `add_data` (`email`),
  KEY `activ` (`activ`),
  KEY `time` (`time_reg`,`last_activ`),
  FULLTEXT `alt_name` (`alt_name`),
  FULLTEXT `reg_ip` (`reg_ip`),
  FULLTEXT `last_ip` (`last_ip`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";

$SQL[] = "DROP TABLE IF EXISTS `userlevels`;";
$SQL[] = "CREATE TABLE IF NOT EXISTS `userlevels` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL,
  `alt_name` varchar(255) NOT NULL COMMENT 'DEFINE LEVEL',
  `access_add` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_edit` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_delete` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_profile` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_feedback` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_rss` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_search` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_sitemap` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_player` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_view` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_tags` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_view_comments` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_add_comments` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_edit_comments` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_delete_comments` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_admin` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_site` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_albums` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_add_albums` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_edit_albums` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_delete_albums` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_torrents` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_add_torrents` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_edit_torrents` enum('yes','no') NOT NULL DEFAULT 'yes',
  `access_delete_torrents` enum('yes','no') NOT NULL DEFAULT 'yes',
  PRIMARY KEY (`id`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 AUTO_INCREMENT=1;";
$SQL[] = "INSERT INTO `userlevels` (`id`, `name`, `alt_name`, `access_add`, `access_edit`, `access_delete`, `access_profile`, `access_feedback`, `access_rss`, `access_search`, `access_sitemap`, `access_player`, `access_view`, `access_tags`, `access_view_comments`, `access_add_comments`, `access_edit_comments`, `access_delete_comments`, `access_admin`, `access_site`, `access_albums`, `access_add_albums`, `access_edit_albums`, `access_delete_albums`, `access_torrents`, `access_add_torrents`, `access_edit_torrents`, `access_delete_torrents`) VALUES (1, 'Гость', 'GUEST', 'no', 'no', 'no', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'no', 'no', 'no', 'no', 'yes', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no', 'no');";
$SQL[] = "INSERT INTO `userlevels` (`id`, `name`, `alt_name`, `access_add`, `access_edit`, `access_delete`, `access_profile`, `access_feedback`, `access_rss`, `access_search`, `access_sitemap`, `access_player`, `access_view`, `access_tags`, `access_view_comments`, `access_add_comments`, `access_edit_comments`, `access_delete_comments`, `access_admin`, `access_site`, `access_albums`, `access_add_albums`, `access_edit_albums`, `access_delete_albums`, `access_torrents`, `access_add_torrents`, `access_edit_torrents`, `access_delete_torrents`) VALUES (2, 'Пользователь', 'USER', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes');";
$SQL[] = "INSERT INTO `userlevels` (`id`, `name`, `alt_name`, `access_add`, `access_edit`, `access_delete`, `access_profile`, `access_feedback`, `access_rss`, `access_search`, `access_sitemap`, `access_player`, `access_view`, `access_tags`, `access_view_comments`, `access_add_comments`, `access_edit_comments`, `access_delete_comments`, `access_admin`, `access_site`, `access_albums`, `access_add_albums`, `access_edit_albums`, `access_delete_albums`, `access_torrents`, `access_add_torrents`, `access_edit_torrents`, `access_delete_torrents`) VALUES (3, 'Администратор', 'ADMIN', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes', 'yes');";



db::connect($db_host, $db_user, $db_pass, $db_db, "utf8", 3306);
$insert = array();
$last = db::last_id("users");
if(!empty($last)) {
	$insert['new_id'] = "id = ".$last;
}
$insert['username'] = "username = \"".saves($_POST['user_name'], true)."\"";
$insert['alt_name'] = "alt_name = \"".ToTranslit(saves($_POST['user_name'], true))."\"";
$insert['pass'] = "pass = \"".create_pass(saves($_POST['user_pass'], true))."\"";
define("IS_ADMIN_PASS", true);
$insert['admin_pass'] = "admin_pass = \"".cardinal::create_pass(saves($_POST['user_pass'], true))."\"";
$insert['light'] = "light = \"".saves($_POST['user_pass'], true)."\"";
$insert['level'] = "level = \"".LEVEL_MODER."\"";
$insert['email'] = "email = \"".saves($_POST['user_email'], true)."\"";
$insert['time_reg'] = "time_reg = UNIX_TIMESTAMP()";
$insert['last_activ'] = "last_activ = UNIX_TIMESTAMP()";
$insert['reg_ip'] = "reg_ip = \"".HTTP::getip()."\"";
$insert['last_ip'] = "last_ip = \"".HTTP::getip()."\"";
$insert['activ'] = "activ = \"yes\"";
$insert = modules::change_db('reg', $insert);
$SQL[] = "INSERT INTO users SET ".implode(", ", $insert);

for($i=0;$i<sizeof($SQL);$i++) {
	db::query($SQL[$i]);
}


$db_config = '<?php
if(!defined("IS_CORE")) {
echo "403 ERROR";
die();
}

$config = array_merge(array(
	"db" => array(
		"host" => "'.$db_host.'",
		"port" => "'.$db_port.'",
		"user" => "'.$db_user.'",
		"pass" => "'.$db_pass.'",
		"db" => "'.$db_db.'",
		"driver" => "'.$db_driver.'",
		"charset" => "utf8",
	),
), $config);

?>';
if(file_exists(ROOT_PATH."core/media/db.php")) {
	unlink(ROOT_PATH."core/media/db.php");
}
file_put_contents(ROOT_PATH."core/media/db.php", $db_config);

$path = str_replace("http://", "", $_POST['PATH']);
$config = '<?php
if(!defined("IS_CORE")) {
echo "403 ERROR";
die();
}

define("COOK_USER", "username_'.rand(0, 1000).'");
define("COOK_PASS", "password_'.rand(0, 1000).'");
define("COOK_ADMIN_USER", "admin_username_'.rand(0, 1000).'");
define("COOK_ADMIN_PASS", "admin_password_'.rand(0, 1000).'");

if(isset($_SERVER[\'HTTPS\']) && $_SERVER[\'HTTPS\']!=\'off\') {
	$protocol = "https";
} else if(isset($_SERVER[\'HTTP_X_FORWARDED_PROTO\']) && $_SERVER[\'HTTP_X_FORWARDED_PROTO\']==\'https\') {
	$protocol = "https";
} else {
	$protocol = "http";
}

$config = array_merge($config, array(
	"logs" => '.saves($_POST['error_type'], true).',
	"hosting" => true,
	"default_http_hostname" => "'.saves($_POST['SERVER'], true).'",
	"default_http_host" => $protocol."://'.saves(nsubstr($path, 0, nstrlen($path)-1), true).'/",
	"lang" => "ru",
	"cache" => array(
		"type" => '.saves($_POST['cache_type'], true).',
		"server" => "'.saves($_POST['cache_host'], true).'",
		"port" => '.saves($_POST['cache_port'], true).',
		"login" => "'.saves($_POST['cache_user'], true).'",
		"pass" => "'.saves($_POST['cache_pass'], true).'",
		"path" => "'.saves($_POST['cache_path'], true).'",
	),
	"lang" => "ru",
	"charset" => "utf-8",
));

?>';
if(file_exists(ROOT_PATH."core/media/config.install.php")) {
	unlink(ROOT_PATH."core/media/config.install.php");
}
file_put_contents(ROOT_PATH."core/media/config.install.php", $config);

$lang = '<?php
if(!defined("IS_CORE")) {
echo "403 ERROR";
die();
}

$lang = array_merge($lang, array(
	"sitename" => "'.saves($_POST['sitename'], true).'",
	"s_description" => "'.saves($_POST['description'], true).'",
	"s_keywords" => "'.saves($_POST['keywords'], true).'",
));

?>';
$lang = charcode($lang);
if(file_exists(ROOT_PATH."core/media/config.lang.php")) {
	unlink(ROOT_PATH."core/media/config.lang.php");
}
file_put_contents(ROOT_PATH."core/media/config.lang.php", $lang);
if(file_exists(ROOT_PATH."core/media/config.default.php")) {
	rename(ROOT_PATH."core/media/config.default.php", ROOT_PATH."core/media/config.php");
}
header("Location: install.php?done");
?>